### ISSUE TITLE
excessive JS/DOM element allocations causing app wide performance degredation

### ISSUE THREAD

**Issue by larkin** [nasa/openmct#732]
I've been digging into "performance degradation while editing" and been able to lock down the issue as unrelated to editing.  However, it is possible the issue is exacerbated by changes made to mct-representation to support new edit mode.

Notes from my research:

If you have two large parent tree nodes (say, 100 elements per), and you expand them both (to load composition) and then close them, you'll drastically spike the tree watch count (in testing, I can easily get the tree to 8k watches).  This significantly slows down digests.

Once you've done this, every time you navigate to one of the parent nodes (can just navigate back and forth), the ng-repeat re-builds all of the nodes contained in the node.  This creates a ton of DOM elements and seems to trigger excessive refreshes of mct-representation.

Here's a resource timeline that I captured with a low (~800) watch count, while toggling between two folders:

![switching between channel nodes that have not been expanded](https://cloud.githubusercontent.com/assets/226503/13657350/7eb8bf04-e623-11e5-872a-690c4d132560.png)

You can see significant allocations, but the GC is running and freeing up space.

It seems that when there is significant processing to be done (e.g. when you've got 8k watches), the GC doesn't run frequently and chrome starts eating up memory.  At some point, performance goes through the floor and the app is basically unusable.

Case in point: With about 8k active watches, Here's a resource timeline where I started selecting one node, selected another node, and then selected the original node.  Each node had ~100 children.  The nodes themselves did not have a main view, which leads me to believe that this is due to tree resource allocation (though that's not certain).  You can see that node count and js heap size both increase per interaction.

![screen shot 2016-03-09 at 6 15 25 pm](https://cloud.githubusercontent.com/assets/226503/13657326/363f5e5e-e623-11e5-9ab3-1a92add022fb.png)

At the same time, GCs don't run even though significant allocations have occured.  It seems that chrome is happy to consume large amounts of memory (I can get my JS heap to ~3gb easily).  I'd expect the V8 GC to be pretty aggressive but it's not even running.  Manually triggering a GC does free up most of the resources.

Another way to reproduce this: in an environment with a lot of searchable objects, search for "a", and then "a " (a followed by a space).  The same results will be shown, and you can keep typing spaces and each space will recreate the search results and thrash a bunch of elements.

A few things could be fixed:
* we should use "track by" domainObjectId (or index) in ng-repeats to reduce re-allocation of dom objects.
* we should reduce watch count in the tree to give the GC more time to run (i think?) and generally improve performance.
* we should prevent MCTRepresentation from calling `refresh` three times for every time it is linked. (two $watch initializations call it, and then it's called at the end of the link function).
* we should simplify the tree markup (reduce the number of DOM elements per node).

@VWoeltjen I've assigned to you because I wanted to get your input.  We could merge the mct-refresh and mct-observe changes which would significantly reduce watch counts and would reduce the scale of the issue, and is probably a faster solution than waiting to replace the tree with a non-angular tool.  But have to look at all of these problems together, too.

Bottom line, our js memory usage is absurdly high.  I'm amazed how huge the heap gets in a short period of time.

**Comment by VWoeltjen** [comment]
This is well-researched - thanks. I'd like to look at how a few solutions (e.g. mct-refresh/mct-observe, simple track-by-index, etc) effect these metrics and compare that against the risk/scope of changes. Replacing the tree with something non-Angular is possible before the April delivery (and could inform some aspects of the API redesign) but that's an option of last resort.

**Comment by larkin** [comment]
@VWoeltjen something that struck me about 2:10am last night: If there is any bug in `unchanged` that results in unnecessary refreshes being done, this would be problematic.  Due to multiple calls to `refresh` during link and the explicit destruction of representations, nested representations could have a 3x multiplier on $scope and dom element creation.

**Comment by VWoeltjen** [comment]
Did a first-pass looking at a few different branches, trying to assess impact on number of nodes created; looking at "Peak Nodes" as the simple metric here (how many total nodes were allocated?). Test case is to start with My Items expanded/selected, click two different folders twice each (to navigate), then navigating back to My Items. Typically spending about 10s on this.

Branches tried on:

* `master`: Establish a baseline
* `open315`: Work-in-progress branch for #315 (many major regressions observed in this branch in its current state)
* `tree-track-by`: Branch from @larkin adding a `track by` clause to `ng-repeat`s in tree, tracking by domain object identifier.

Branch | Peak Nodes | Risk | Effort
------|-------|--------|-----
`master` | 35k | N/A | N/A
`open315` | 68k | High | Moderate
`tree-track-by` | 38k | Low | Low

> @VWoeltjen something that struck me about 2:10am last night: If there is any bug in unchanged that results in unnecessary refreshes being done, this would be problematic. Due to multiple calls to refresh during link and the explicit destruction of representations, nested representations could have a 3x multiplier on $scope and dom element creation. 

Around 2:10am last night I realized that I had a final tomorrow and had forgotten to attend classes _all semester._ You're doing a little better than me.

This rings a bell - I believe there was an issue where this was encountered and (partially?) resolved. I believe it was in the context of plots. I've done some issue-searching but haven't encountered any notes on this. (@akhenry, does this sound familiar to you at all?)

**Comment by larkin** [comment]
@vwoeltjen there was a plot issue with scope reuse where plot controllers were sticking around and every navigation change would trigger the number of telemetry requests to increase. I think that is what led to the changes to mct representation to prevent scope reuse.

**Comment by larkin** [comment]
@vwoeltjen if you want a more stressing case try running warp and doing the same with some of the larger packets. If the watch count is too low and there is lots of idle time then chrome will garbage collect more proactively. When there is more processing occurring it won't garbage collect until t absolutely must, and this causes the performance to degrade.

**Comment by VWoeltjen** [comment]
Ah, thank you, you are correct!

Added some metrics-reporting to `mct-representation` (see branch `metrics-732`) to report time an `mct-representation` gets linked, how many `refresh` calls are made, how many of those calls get past the `unchanged` check, and how many further passed the `canRepresent` check:

* Every node in the tree reports that 3 `refresh` calls were made and that all three calls got past the `canRepresent` check. So, the template for each representation is getting reloaded (and relinked) a full three times. This does make it seem that something has gone wrong in the `unchanged` check (but, this has all happened and been forgotten by the time we start navigating)
* The reported timestamp for the `link` call for each node does _not_ change during navigation, suggesting that nodes are not being recreated during navigation.

My test case in this instance was using the `example/taxonomy` plugin, expanding Stub Subsystem 1 and 2, and navigating back and forth between the two. 

@larkin wrote:

> Once you've done this, every time you navigate to one of the parent nodes (can just navigate back and forth), the ng-repeat re-builds all of the nodes contained in the node. This creates a ton of DOM elements and seems to trigger excessive refreshes of mct-representation.

How were you able to observe this? I'm trying to verify that this (and not something else) accounts for the increase in allocated nodes, but haven't found any direct evidence.

**Comment by VWoeltjen** [comment]
Repeating my test case from above (using `example/taxonomy`) and disabling Inspector and the view-object area (via `ng-if="::false"`) I observe:

Inspector Present | View-object Present | Nodes per click (approx)
:---:|:---:|:---:
:white_check_mark: | :white_check_mark: |
:no_entry_sign:  | :white_check_mark: |
:white_check_mark: | :no_entry_sign: | 22000
:no_entry_sign: | :no_entry_sign: | 300

**Comment by larkin** [comment]
@VWoeltjen i wrote a little later on ("Case in point") that I couldn't verify that this-- and not something else-- was causing it.  But I imagine it would be possible to comment out other representations so that only the tree is visible and that should isolate it.

**Comment by VWoeltjen** [comment]
Possibly - at least, the allocation patterns look a lot different with Inspector suppressed. Could be a more complex interaction that happens to involve the Inspector.

Also want to try with a Folder instead of a Plot view - curious how snappy that is (expect a lot of allocations for the Items view of a Folder)

**Comment by larkin** [comment]
I tested folders with user folders and it did allocate but when I was testing I had a pretty low watch count and it was able to garbage collect.

![switching between folders](https://cloud.githubusercontent.com/assets/226503/13685110/c58906fa-e6c3-11e5-8e1a-64c9634828cf.png)

**Comment by VWoeltjen** [comment]
Node allocation jumps to 70-100k with a Folder instead of a Plot view showing. (Not entirely unexpected, as Items view does allocate O(n) nodes for n elements in composition; still, the number of actual node allocations seems exceptionally high.)

**Comment by VWoeltjen** [comment]
Spotted the bug in the `unchanged` check - it's obvious once you see it (see daf41f1, which I'm guessing we'll want to cherry-pick later.)

With this change in place, representations in the tree and elsewhere report only 1 actual refresh (but are still checked 3 times, which should be fine), and node allocations in the case of the `view-object` being suppressed (just tree + inspector showing) drop from 22000 to 9000 per navigation.

I still think we'll have problems with memory usage when digest cycles are frequent enough that Chrome decides not to bother GCing, but the fix to `unchanged` should slow the rate at which that occurs significantly.

The other change to consider would be to use _some_ of the changes for #315 in the tree, but not all. Getting the tree watches down to zero is necessary if we want to get the watch count below 2000 (because the number of visible nodes is easily in the 100s, and the rest of the UI needs a decent watch budget too), but cutting down on digests there could still have a noticeable impact on performance & memory usage in these cases.

**Comment by larkin** [comment]
@VWoeltjen ah good catch.

Same time, I agree that 7k nodes seems pretty outrageous.

**Comment by VWoeltjen** [comment]
With this change, still do see the issues with an apparently-preempted GC in circumstances similar to original issue.

Specifically, testing with `example/taxonomy`, two large nodes expanded (>500 total objects visible in tree), and clicking back and forth between two empty folders. JS Heap rises about 10MB per-click, while nodes increase by about 1100 per click. Subjectively, performance seems fairly responsive.

What's peculiar is that metrics differ radically depending on which of the Chrome Developer Tools I use:

* Capturing Memory from the Timeline reports JS Heap usage as growing from 717 to 747 MB over the course of 10 clicks (there does appear to be some GC occurring.) 
* Profiles > Take Heap Snapshot gives me a snapshot with 45-49 MB of memory being used. (Similar results if I Record Heap Allocations instead.)

I'm rather confounded by this - I'd like to use the latter tool to gain a better understanding of the numbers I'm seeing in the former, but they're completely mismatched. Will read through Chrome documentation to see if I can make more sense of this.

> Same time, I agree that 7k nodes seems pretty outrageous.

It is, as is the total node count for the application (50-70k). By comparison, https://github.com/nasa/openmctweb has 3400 nodes allocated during normal use. Our UI does have more complexity, but by a factor of 20?

IIRC every attribute in the DOM (as well as every element, comment, etc) is a node, so this number might be inflated by the number of Angular directives. Would like to have a better understanding of what these allocations really are, and where/when they are occurring. The `unchanged` bug may have had a multiplier effect for nested `mct-representation`s, wondering if anything else might also be doing this.

**Comment by larkin** [comment]
@VWoeltjen ran into this same problem with differences between heap snapshots (resources tab) and the timeline tab's js heap.  My understanding is that heap snapshots and heap allocations will run a complete and aggressive garbage collect before capturing-- they will show you true memory leaks where the GC cannot free memory because a reference is being held.

The timeline will show growth over time -- without forcing garbage collections; it allows chrome to GC as it likes.  And the issue here is chrome is happy to allow a tab to use gigabytes of memory if that tab is also doing heavy processing.  When we do eventually hit the memory limit, chrome is forced to GC before any new allocations can occur and this is slow.  Or perhaps chrome is never GCing, and it's just using virtual memory and that is slow.

**Comment by larkin** [comment]
@VWoeltjen another thing that allocates memory is defensive programming that uses object allocations, e.g.:

`var uses = ((representation || {}).uses || [])`.  If representation is undefined, this allocates an object and an array which are thrown away, when in truth, if representation is undefined, we are going down a different code path anyways and don't care about what `uses` is.

Our code is littered with these sorts of defensive statements.  This is one thing about 'singular var declaration' that I absolutely hate: it encourages you to do all your declarations and assignments up front even if the logical flow of the application would result in it being unnecessary.  

An Aside: Bleh; I hate combined var statements.  YES, I know-- variable hoisting is a thing, I'm a developer and I understand that, but I consider this style a total antipattern that makes code flow harder to understand and encourages this poor method of defensive programming and most importantly, does not prevent you from accidentally double assigning or reassigning a variable at a later date.

**Comment by akhenry** [comment]
Combined var statements feel a little weird to me. I'm used to declaring short-use variables with limited 'scope' (air quotes - variable hoisting etc.) close to where they're used, which makes things a little easier to follow. But it also provides motivation to keep functions small :)

**Comment by VWoeltjen** [comment]
For a second I thought I'd observed memory usage go down when running a built version. I didn't, but when I went back and did a "fair fight" (starting from a fresh page load in both cases) I saw memory usage reported by Timeline drop quite a bit from what I'd seen previously.

Now, with >600 nodes showing in the tree, I see memory regularly returning to 50MB (and not exceeding 100MB). I'm repeating the same steps where I regularly saw Timeline report 650-750MB and can't get it anywhere close to those numbers.

__Hypothesis__: There's some quirk (bug?) in Chrome (either the dev tools or the GC) that was causing memory retained from before the fix from #734 to stick around and show up when profiling _after_ the changes for #734. (For this hypothesis to be correct, such a quirk/bug must not effect node counts, since I _did_ observe those change.) 

For purposes of data collection, my not-very-interesting built-versus-unbuilt data:

------

Steps (running master, post merge of #734 with `example/taxonomy`):

* Load application, wait for fade effect to end
* Open Chrome Developer Tools
* Start recording in Timeline with Memory selected
* Expand all Stub Subsystems (leaving navigation on My Items)
* Navigate to an empty folder 4 times (I have two empty folders side by side to facilitate this)
* End recording in Timeline
* Go to Profile tab and take a Heap Snapshot

Heap sizes below in MB

Built? | Low Heap | High Heap | Final Heap | Low Nodes | High Nodes | Final Nodes | Heap Snapshot Size
:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:
No   | 19  | 79  | 70  | 10k | 59k | 59k | 49
Yes  | 18  | 80  | 72  | 7k  | 57k | 57k | 49

**Comment by VWoeltjen** [comment]
Hm, so now I can't reproduce the numbers I was originally seeing at all.

I am able to get memory usage up to 200-300MB by going back to ff5f37d (before the fix to `unchanged`), then observed numbers still in this range when checking out 5296255 (after the fix to `unchanged`) and reloading. Opening a new tab, memory usage goes back down to 50-150MB repeating the same test scenario (and stays in this range even if I reload and repeat). This is with >600 tree nodes visible.

I'd like to continue looking at this issue, but based on these observations I'd like to shift focus away from memory utilization and toward node count specifically (those numbers just look exceptionally high)

**Comment by larkin** [comment]
The big thing is having enough processing to do that v8 decides not to GC. I imagine I can still cause that to occur in production environments due to large trees and watch counts. 

Large node counts or memory allocs is bad, but if watch counts are down, there is more time to GC, and the problem doesn't get out of hand. 

I think reducing watches would be our best investment right now. 

Moving away from angular and loading all of our ui concerns into the Dom (directives) will be a good way to get node counts down. 

Similarly, getting away from scopes and their non deterministic logic flows will reduce the amount of defensive programming we do and reduce JS allocs.

**Comment by VWoeltjen** [comment]
> @VWoeltjen another thing that allocates memory is defensive programming that uses object allocations, e.g.:

> `var uses = ((representation || {}).uses || [])`.  If representation is undefined, this allocates an object and an array which are thrown away, when in truth, if representation is undefined, we are going down a different code path anyways and don't care about what `uses` is.

Do agree with this point, another reason to move away from that style. (On the other hand, "this object won't be used when we go down a certain code path" is the sort of thing I'd expect an optimizer to figure out...)

**Comment by charlesh88** [comment]
Sprint review 3/29/16: this is fixed, closing.
